### ISSUE TITLE
colcon test-result 'skipped' count is inaccurate

### ISSUE THREAD

**Issue by pbaughman** [colcon/colcon-test-result#18]
When you run colcon test-result you get a little summary of all the tests run:

```
pete.baughman@ade:~/ws$ colcon test-result
Summary: 7644 tests, 0 errors, 4 failures, 1 skipped
```

The 'skipped' test field misses pytests skipped with the `@skip` decorator, and gtests skipped with the `GTEST_SKIP` macro.

The 1 skip in the above report is from test_interfaces_cpp.gtest.xml.  You can see from the XML, this is actually a 'disabled' test
```
<testsuites tests="21" failures="0" disabled="1" errors="0" timestamp="2019-04-22T12:32:50" time="0" name="AllTests">
  <testsuite name="Test_rosidl_generator_traits" tests="2" failures="0" disabled="0" errors="0" time="0">
    <testcase name="has_fixed_size" status="run" time="0" classname="rosidl_generator_cpp.Test_rosidl_generator_traits"/>
    <testcase name="has_bounded_size" status="run" time="0" classname="rosidl_generator_cpp.Test_rosidl_generator_traits"/>
  </testsuite>
  <testsuite name="Test_messages" tests="19" failures="0" disabled="1" errors="0" time="0">
    <testcase name="primitives_static" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="primitives_static_arrays" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="primitives_bounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="primitives_unbounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="static_array_static" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="static_array_bounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="static_array_unbounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="bounded_array_static" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="bounded_array_bounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="bounded_array_unbounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="unbounded_array_static" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="unbounded_array_bounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="unbounded_array_unbounded" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="primitives_constants" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="primitives_default" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="string_arrays_default" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="DISABLED_Test_bounded_strings" status="notrun" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="Test_string" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
    <testcase name="Test_string_array_static" status="run" time="0" classname="rosidl_generator_cpp.Test_messages"/>
  </testsuite>
</testsuites>
```

This particular test is skipped by prefixing the name with DISABLED_
from rosidl_generator_cpp/test/test_interfaces.cpp
```cpp
 // TODO(mikaelarguedas) reenable this test when bounded strings enforce length
TEST(Test_messages, DISABLED_Test_bounded_strings) {
  rosidl_generator_cpp::msg::StringBounded message;
  TEST_STRING_FIELD_ASSIGNMENT(message, string_value, "", "Deep into")
  std::string tooLongString = std::string("Too long string");
  message.string_value = tooLongString;
  tooLongString.resize(BOUNDED_STRING_LENGTH);
  ASSERT_STREQ(tooLongString.c_str(), message.string_value.c_str());
}
```
Skipped pytests have XML that looks like this:
```
<?xml version="1.0" encoding="utf-8"?>
<testsuite errors="0" failures="0" name="pytest" skipped="6" tests="175" time="40.009">
  . . .
  <testcase classname="rclpy.test.test_timer" file="test/test_timer.py" line="150" name="test_timer_zero_callbacks100hertz" time="0.001">
    <skipped message="&lt;Skipped instance&gt;" type="pytest.skip">/usr/local/lib/python3.5/dist-packages/_pytest/nose.py:29: &lt;Skipped instance&gt;</skipped>
  </testcase>
  . . .
</testsuite>
```

## Suggested Fix
Tests that are skipped with pytest @unittest.skip have an additional child element under the skipped testcase that says the skipped reason.  I have a tool that contains logic sort of like this:

```python
# Things get a little messy down at the test case level.  Some test cases have a "status"
# attribute we can inspect - but it's optional
if 'status' in case.attrib:
    status = case.attrib['status']
    if status == 'notrun' or status == 'skipped':
        return True
# Pytest XML has a 'skipped' element inside the test case
if case.findall('skipped'):
    return True
return False
```

We could also look for the `skipped` attribute on the `testsuite` element and only drill down to the test cases if it's not present.

**Comment by pbaughman** [comment]
I intend to fix this myself as soon as I have a free second.  Just leaving this here so I remember.  Feedback welcome if you feel I've mis-identified the problem.

**Comment by dirk-thomas** [comment]
Thank you for filling this ticket. There are certainly too many variations of jUnit-like result files out there in the wild :disappointed:

The code already tries to look for various possible attributes: https://github.com/colcon/colcon-test-result/blob/3e00b0859c14e8a74b7da81086bfa46da2fe60fd/colcon_test_result/verb/test_result.py#L187-L189

As you can see from that snippet disabled tests are also considered "skipped" for the sake of keeping the number of categories low.

Injecting the following line (in the middle) should likely resolve the problem:

```
('skipped_count', 'skipped', 0),
```

If you could give it a try and report back (or even better create a PR with it) that would be great.

**Comment by pbaughman** [comment]
@dirk-thomas Yeah, I'll try to slam out a PR before the end of the week - thanks for the tip

**Comment by pbaughman** [comment]
@dirk-thomas - ok, that picks up the skips from pytests (woo!)

gtests skipped with the GTEST_SKIP(); macro are another story.  This is arguably a bug in googletest but the XML for a skipped test case looks like this:
```
<testsuites tests="1" failures="0" disabled="0" errors="0" timestamp="2019-04-25T13:13:50" time="0.022" name="AllTests">
  <testsuite name="TestExecutors" tests="1" failures="0" disabled="0" errors="0" time="0.016">
    <testcase name="detachOnDestruction" status="skipped" time="0.016" classname="rclcpp.TestExecutors">
    . . .
```
As you can see - the status of the `testcase` is set to 'skipped' but there's no mention of it at the `testsuite` or `testsuites` level

Just for fun, I updated to the latest upstream master of googletest (ahead of ament/googletest) and the XML looks like this
```
<testsuites tests="1" failures="0" disabled="0" errors="0" timestamp="2019-04-25T13:38:00" time="0.016" name="AllTests">
  <testsuite name="TestExecutors" tests="1" failures="0" disabled="0" errors="0" time="0.011">
    <testcase name="detachOnDestruction" status="run" result="skipped" time="0.011" classname="rclcpp.TestExecutors"/>
```
'status' now says "run" instead of "skipped" and there's a new result attribute set to "skipped"  Still no mention of the skips at the `testsuites` or `testsuite` level

#### In Conclusion:
Does it bother you if the 'skipped' count doesn't work for tests skipped with GTEST_SKIP()?  I opened this can of worms here: https://github.com/ament/googletest/pull/4 (sorry!) and am fairly certain nobody else is using the GTEST_SKIP macro to skip a gtest.

I think the 'right' fix is to have googletest write a skipped count to the testsuite element.

We could also "fix" it in colcon test-result by drilling down to the individual `testcase` elements when no skipped attribute is found on the testsuite element.

I'm also happy to leave it alone and just submit a fix for skipped pytests - which [actually exist in ROS](https://github.com/ros2/rclpy/blob/616e853455bfd236e61ef09f92b4c5afe15af389/rclpy/test/test_timer.py#L157-L164)

**Comment by dirk-thomas** [comment]
> Does it bother you if the 'skipped' count doesn't work for tests skipped with GTEST_SKIP()?

It would of course be nice if it would be taken into account. But that seems like a problem in `googletest` (see next paragraph).

> I think the 'right' fix is to have googletest write a skipped count to the testsuite element.

That sounds like the right step forward. Did you already open a ticket in the upstream repo for that? If no, can you please do so. If yes, can you link to it here for future readers.

**Comment by pbaughman** [comment]
@dirk-thomas The commit that fixes googletest can be found here:

https://github.com/pbaughman/googletest/commit/c4f09e05f2d2cc87a260fe9492f8dc57200047d7

Here's a branch with the same commit cherry-picked onto the 'ros2' branch of the ament googletest fork:
https://github.com/pbaughman/googletest/tree/add_skipped_counts_ament_version

The CLA for me to submit a PR to the googletest repo is a bit much for me to tackle right now.  Sorry - I'll reply here if/when I manage to get this upstream.  In the meantime, if anybody else is getting burned by this, they can take the fixes linked above. ^^

**Comment by dirk-thomas** [comment]
I understand the extra steps is annoying. It is not a viable option for us to switch to your fork of `googletest` for this though. So please try to get the change upstream and we are more than happy to sync our fork.

**Comment by dirk-thomas** [comment]
Currently the problem is only mentioned here in a closed ticket. I would encourage you to open an issue on the upstream repo for visibility / awareness.

**Comment by pbaughman** [comment]
Yeah, I get it.  If I get this into googletest I'll update you.

**Comment by dejanpan** [comment]
@dirk-thomas do you guys have CLA signed for googletest:https://github.com/google/googletest/blob/master/CONTRIBUTING.md#contributor-license-agreements?

**Comment by dirk-thomas** [comment]
> do you guys have CLA signed for googletest

Not that I know of.

**Comment by Ryanf55** [comment]
I was curious - is there a recommendation for humble? 
Using `GTEST_SKIP()` results in `colcon test-result` misreporting a skipped test as a passed test.

**Comment by wsxq2** [comment]
> I was curious - is there a recommendation for humble? Using `GTEST_SKIP()` results in `colcon test-result` misreporting a skipped test as a passed test.

I faced the same problem.
